src/routes/publish.ts

```diff
@@ -8,196 +8,602 @@ import { StatsWalesApi } from '../services/stats-wales-api';
+    const lang = req.i18n.language;
+    if (processedCSV.success) {
+        const viewDTO = processedCSV as ViewDTO;
+        if (!viewDTO.dataset) {
+            res.status(500);
+            res.render('publish/upload', processedCSV);
+            return;
+        }
+        setCurrentToSession(viewDTO.dataset, req);
+        res.redirect(
+            `/${lang}/${req.i18n.t('routes.publish.start', { lng: lang })}/${req.i18n.t('routes.publish.preview', { lng: lang })}`
+        );
+    } else {
+        const err = processedCSV as UploadErrDTO;
+        req.session.errors = generateViewErrors(err.dataset?.id, 500, err.errors);
+        res.redirect(`/${lang}/${req.i18n.t('routes.publish.start', { lng: lang })}/`);
+    }
+}
+
+async function createNewDataset(req: AuthedRequest, res: Response) {
+    const lang = req.i18n.language;
+    const statsWalesApi = new StatsWalesApi(lang, req.jwt);
+    const title = req.session.currentTitle;
+    if (!title) {
+        logger.debug('Current title was missing from the session.  Something might have gone wrong');
+        req.session.errors = generateViewErrors(undefined, 400, [generateError('title', 'errors.title_missing', {})]);
+        res.redirect(
+            `/${lang}/${req.i18n.t('routes.publish.start', { lng: lang })}/${t('routes.publish.title', { lng: lang })}`
+        );
+        return;
+    }
+    const file = req.file;
+    if (!file) {
+        generateFileError(req, res);
         return;
     }
 
-    const title: string = req.body.title;
+    const fileName = file.originalname;
+    const fileData = new Blob([file.buffer], { type: file.mimetype });
+    const processedCSV = await statsWalesApi.uploadCSVtoCreateDataset(fileData, fileName, title);
+    await handleProcessedCSV(processedCSV, req, res);
+}
+
+async function uploadNewFileToExistingDataset(req: AuthedRequest, res: Response) {
+    const lang = req.i18n.language;
+    const statsWalesApi = new StatsWalesApi(lang, req.jwt);
+    const currentDataset = checkCurrentDataset(req, res);
+    const currentRevision = checkCurrentRevision(req, res);
+    if (!currentDataset || !currentRevision) {
+        return;
+    }
 
     if (!req.file) {
-        const err: ViewErrDTO = {
-            success: false,
-            status: 400,
-            dataset_id: undefined,
-            errors: [
-                {
-                    field: 'csv',
-                    message: [
-                        {
-                            lang,
-                            message: t('errors.upload.no-csv-data')
-                        }
-                    ],
-                    tag: {
-                        name: 'errors.upload.no-csv-data',
-                        params: {}
-                    }
-                }
-            ]
-        };
-        res.status(400);
-        res.render('publish/upload', err);
+        generateFileError(req, res);
```

Comment:
Because it creates the error and sets the return status and the render.  So if you do this you just need to return... That could be clearer in the function name.  So no return object... just create an error and pass it back to the view.

Reply:
Ah, was getting confused with `generateError`